### ISSUE TITLE
NGP

### ISSUE THREAD

**Issue by JMcKiern** [JMcKiern/warhawk-reversing#1]
# Overview

## Header of size 0x20

|Offset|Length|Example|Name|Notes|
|---|---|---|---|---|
|0x00|0x08|0x696570334616A42B|Magic Number|NGPs always start with this|
|0x08|0x04|0x466F37AB|Object Identifier|Present in PTR and VRAM file headers for same object|
|0x0C|0x04|0x0000001C|Table 1 Pointer|Relative pointer|
|0x10|0x04|0x00000070|Texture Table (Table 2) Pointer|Relative pointer|
|0x14|0x04|0x00095FB4|Table 3 Pointer|Relative pointer|
|0x18|0x04|0x00002F68|Table 4 Pointer|Relative pointer|
|0x1C|0x04|0x000022E0|Table 5 Pointer|Relative pointer|

## Overall File Structure
- Header
- Table 1
- Texture Table
- Unknown
- Model headers (maybe includes pointer to texture headers?)
- Unknown
- Model data
	1. Face data - Each face is 3 vertex indices which are shorts
	2. Vertex data - short_signed
- Unknown
- Snippets of stuff with a header (at the bottom of the data?) that includes
  strings like "hud" and "vehicle"
- Unknown
- NGP Texture headers
- Unknown
- Texture data (pointed to by the above texture headers)
- Unknown

### UV Coordinates

The UV coordinates seem to be centred around 0x3800. However, they're not linear, with the following transformation happening:

|Raw Coordinate Data|Texture Space Coordinate|
|---|---|
|0x17DD|Approaching 0|
|0x3800|0.5|
|0x3C1A|Approaching 1|

The following function seems to work for what I've tested...

```python
def translate(val):
        return (((val / 0x3800) ** 10) / 2)
```

### Standard Table 

Offset | Length | Example | Name | Notes
-- | -- | -- | -- | --
0x00 | 0x04 | 0x00000001 | Number of entries in table |  
0x04 + (n * 0x04) | 0x04 | 0x00000804 | Entry in table | Relative pointer

### Relative Pointer

- Here, a relative pointer is defined as a pointer to that points to a file address relative to the pointer's address
	- Eg. a relative pointer with value 0x10 at location 0x4 would point to 0x14

**Comment by JMcKiern** [comment]
# Table 1 (T1)

T1 is a Standard Table

Entries point to Table 1 Data (T1.D)

## Table 1 Data (T1.D)

Offset | Length | Example | Name | Notes
-- | -- | -- | -- | --
0x00 | 0x04 | 0x08010000 | Magic | 
0x04 | ??? | | | End of structure unknown

**Comment by JMcKiern** [comment]
# Table 2 (Texture Table)

Table 2 is a Standard Table

Table entries are relative pointers to NGP Texture Headers

## NGP Texture Header
- Texture headers (length of 0x10)
	- These are .rtt headers but are missing the first 4 bytes
	- The last 4 bytes gives the offset of the texture data
	- Byte 0x8 (or 0xc after adding the first 4 bytes) indicates where the texture data is
	  located
	  	- 0x00 means in the .vram file
		- 0x01 means in the .ngp file

**Comment by JMcKiern** [comment]
# Table 3 (T3)

Offset | Length | Example | Name | Notes
-- | -- | -- | -- | --
0x00 | 0x02 | 0x0006 | Number of entries in T3.0 table |
0x02 | 0x02 | 0x000d | Number of entries in table | 
0x04 | 0x04 | 0xFFFFFFE4 | Pointer to T3.0 Table | Relative pointer
0x08 + (n * 0x04) | 0x04 | 0xFFFFF87C | Entry in table | Relative pointer to T3.1 data

## Table 3.1 Data (T3.1.D)

**Comment by JMcKiern** [comment]
# Model Headers

Model Headers contain the pointers to the faces, vertices, UV coordinates and a data structure that leads to the textures.

#### Model Header Type 1 - Static Mesh
|Offset|Length|Example|Name|Notes|
|---|---|---|---|---|
|0x00|0x04|0x00000001|Magic Number||
|0x04|0x04|0x0000B114|Relative pointer to unknown data structure|The data structure has links to the texture headers|
|0x08|0x04|0x43780000|*Unknown*|This is likely a float|
|0x0C|0x04|0x437E0000|*Unknown*|This is likely a float|
|0x10|0x04|0x439A0000|*Unknown*|This is likely a float|
|0x14|0x04|0x3C000000|*Unknown*|This is likely a float|
|0x18|0x04|0x00000004|*Unknown*||
|0x1C|0x04|0x00000366|Indices Used in Faces|Number of Faces * 3|
|0x20|0x04|0x000001B9|*Unknown*|Is generally Number of Vertices. However, does not always match. Need to investigate.|
|0x24|0x02|0x01B9|Number of Vertices||
|0x26|0x01|0x05|Number of Linkers|headerSize = 0x2C + (numLinkers * 0x0C)|
|0x27|0x01|0x01|*Unknown*||
|0x28|0x04|0x00001F00|Face Offset||
|0x2C|0x04|0x00000005|*Unknown*||
|0x30|0x04|0x06030100|*Unknown*||
|0x34|0x04|0x000025D0|Vertex Offset||
|0x38|0x0C|0x000200041003000000114980|*Unknown Structure*|See below|
|0x44|0x0C|0x000A00041004000000114983|*Unknown Structure*|See below|
|0x50|0x0C|0x000800031002000000114988|*Unknown Structure*|See below|
|0x5C|0x0C|0x00090003100200000011498C|*Unknown Structure*|See below|

These *Unknown Structures* are 96 bit long data structures that link currently unknown data to the model header. They are structured as follows:

|Offset|Length|Example|Name|Notes|
|---|---|---|---|---|
|0x00|0x04|0x00020004|Identifier?|0x00080003 - UV Coordinates (see UV below)<br>0x00090003 - Bump Texture UV Coordinates?|
|0x04|0x01|0x10|Inter-value Distance|The distance between two consecutive values|
|0x05|0x01|0x03|*Unknown*||
|0x06|0x01|0x00|Data Location Flag|- 0x00 - Data in VRAM<br>- 0x01 - Data in NGP|
|0x07|0x01|0x00|*Unknown*||
|0x08|0x04|0x00114980|Data Offset|Absolute address of the data|

#### Model Header Type 2 - Rigged Mesh
|Offset|Length|Example|Name|Notes|
|---|---|---|---|---|
|0x00|0x04|0x00000002|Magic Number||
|0x24|0x04|0x00023B1C|Vertex Relative Pointer||
|0x44|0x04|0x00028100|Face Offset||

**Comment by JMcKiern** [comment]
Closed by f098248e1d735d11ac400a347c6d22b11e5b0656
